src/shims/unix/linux/epoll.rs

```diff
@@ -294,27 +296,28 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_ctl`"))?;
 
         let binding = epoll_file_description.return_ready_list()?;
-        let ready_list = binding.borrow_mut();
+        let mut ready_list = binding.borrow_mut();
         let num_of_events: i32 = 0;
         let mut array_iter = this.project_array_fields(&event)?;
-        // TODO: filter out all the file description that is fully closed.
-        // then it is possible to calculate the actual values and call
-        // and use take for iter
-        for (epoll_key, epoll_return) in ready_list.iter() {
-            if epoll_key.0.get_file_description().is_some() {
+
+        while let Some((epoll_key, epoll_return)) = ready_list.pop_first() {
+            if epoll_key.0.upgrade().is_some() {
                 // Return notification to the caller if the file description is not fully closed.
-                let des = array_iter.next(this)?.unwrap().1;
-                this.write_int_fields_named(
-                    &[("events", epoll_return.events.into()), ("u64", epoll_return.data.into())],
-                    &des,
-                )?;
+                if let Some(des) = array_iter.next(this)? {
+                    this.write_int_fields_named(
+                        &[
+                            ("events", epoll_return.events.into()),
+                            ("u64", epoll_return.data.into()),
+                        ],
+                        &des.1,
+                    )?;
+                }
+                // Terminate early because we only return at most maxevents number of epoll_event.
                 num_of_events.checked_add(1).unwrap();
                 if num_of_events == maxevents {
```

Comment:
This check is now redundant

Reply:
Yup, the check can be omitted. This is unrelated to this check, but I wonder if  ``Ty::new_array`` would handle this case as expected:
1. if size of array passed in < maxevents, would it be terminated earlier by having ``None`` on next or it might potentially leads to OOB memory access?